File: desing-pattens-1/src/teste/OrcamentoTeste02.java
package teste;

import org.junit.Assert;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.jupiter.api.BeforeAll;

import interfaces.IDesconto;
import modelo.Item;
import modelo.Orcamento;
import modelo.implementacao.desconto.DescontoMaisDeCincoItens;
import modelo.implementacao.desconto.DescontoOrcamentoMaiorQue500;
import modelo.implementacao.desconto.DescontoVendaComLapisECaneta;
import modelo.implementacao.desconto.SemDesconto;

public class OrcamentoTeste02 {

	private static IDesconto descontoMaisDeCincoItens;

	private static IDesconto descontoOrcamentoMaiorQue500;

	private Orcamento orcamento;

	private static IDesconto semDesconto;

	private static IDesconto descontoVendaComLapisECaneta;

	@BeforeClass
	public static void configurar() {

		descontoMaisDeCincoItens = new DescontoMaisDeCincoItens();
		descontoOrcamentoMaiorQue500 = new DescontoOrcamentoMaiorQue500();
		descontoVendaComLapisECaneta = new DescontoVendaComLapisECaneta();
		semDesconto = new SemDesconto();
		descontoMaisDeCincoItens.setProximo(descontoOrcamentoMaiorQue500);
		descontoOrcamentoMaiorQue500.setProximo(descontoVendaComLapisECaneta);
		descontoVendaComLapisECaneta.setProximo(semDesconto);
	}

	@Test
	public void testaDecontoParaMaisDe5Itens() {

		orcamento = new Orcamento(100);
		orcamento.adicionar(new Item("CADERNO"));
		orcamento.adicionar(new Item("LIVRO"));
		orcamento.adicionar(new Item("BORRACHA"));
		orcamento.adicionar(new Item("CANETA"));
		orcamento.adicionar(new Item("PINCEL"));
		orcamento.adicionar(new Item("GIZ"));
		Assert.assertTrue(0.1 == descontoMaisDeCincoItens.calcula(orcamento));
	}

	@Test
	public void testaDecontoParaOrcamentoMaiorQue500() {

		orcamento = new Orcamento(1000);
		orcamento.adicionar(new Item("CADERNO"));
		orcamento.adicionar(new Item("LIVRO"));

		Assert.assertTrue(0.07 == descontoMaisDeCincoItens.calcula(orcamento));
	}

	@Test
	public void testaSemDesconto() {

		orcamento = new Orcamento(10);
		orcamento.adicionar(new Item("CADERNO"));
		orcamento.adicionar(new Item("LIVRO"));

		Assert.assertTrue(0 == descontoMaisDeCincoItens.calcula(orcamento));
	}

	@Test
	public void testaDescontoVendaCasada() {

		orcamento = new Orcamento(500);
		orcamento.adicionar(new Item("LAPIS"));
		orcamento.adicionar(new Item("CANETA"));

		Assert.assertTrue(0.05 == descontoMaisDeCincoItens.calcula(orcamento));
	}

}

File: desing-pattens-1/src/teste/teste/OrcamentoTeste02.kt
package teste

import main.java.interfaces.IDesconto
import main.java.modelo.Item
import main.java.modelo.Orcamento
import main.java.modelo.implementacao.desconto.DescontoMaisDeCincoItens
import main.java.modelo.implementacao.desconto.DescontoOrcamentoMaiorQue500
import main.java.modelo.implementacao.desconto.DescontoVendaComLapisECaneta
import main.java.modelo.implementacao.desconto.SemDesconto

class OrcamentoTeste02 {

    private var orcamento: Orcamento? = null

    @Test
    fun testaDecontoParaMaisDe5Itens() {

        orcamento = Orcamento(100.0)
        orcamento!!.adicionar(Item("CADERNO"))
        orcamento!!.adicionar(Item("LIVRO"))
        orcamento!!.adicionar(Item("BORRACHA"))
        orcamento!!.adicionar(Item("CANETA"))
        orcamento!!.adicionar(Item("PINCEL"))
        orcamento!!.adicionar(Item("GIZ"))
        Assert.assertTrue(0.1 == descontoMaisDeCincoItens!!.calcula(orcamento))
    }

    @Test
    fun testaDecontoParaOrcamentoMaiorQue500() {

        orcamento = Orcamento(1000.0)
        orcamento!!.adicionar(Item("CADERNO"))
        orcamento!!.adicionar(Item("LIVRO"))

        Assert.assertTrue(0.07 == descontoMaisDeCincoItens!!.calcula(orcamento))
    }

    @Test
    fun testaSemDesconto() {

        orcamento = Orcamento(10.0)
        orcamento!!.adicionar(Item("CADERNO"))
        orcamento!!.adicionar(Item("LIVRO"))

        Assert.assertTrue(0.0 == descontoMaisDeCincoItens!!.calcula(orcamento))
    }

    @Test
    fun testaDescontoVendaCasada() {

        orcamento = Orcamento(500.0)
        orcamento!!.adicionar(Item("LAPIS"))
        orcamento!!.adicionar(Item("CANETA"))

        Assert.assertTrue(0.05 == descontoMaisDeCincoItens!!.calcula(orcamento))
    }

    companion object {

        private var descontoMaisDeCincoItens: IDesconto? = null

        private var descontoOrcamentoMaiorQue500: IDesconto? = null

        private var semDesconto: IDesconto? = null

        private var descontoVendaComLapisECaneta: IDesconto? = null

        @BeforeClass
        fun configurar() {

            descontoMaisDeCincoItens = DescontoMaisDeCincoItens()
            descontoOrcamentoMaiorQue500 = DescontoOrcamentoMaiorQue500()
            descontoVendaComLapisECaneta = DescontoVendaComLapisECaneta()
            semDesconto = SemDesconto()
            descontoMaisDeCincoItens!!.setProximo(descontoOrcamentoMaiorQue500)
            descontoOrcamentoMaiorQue500!!.setProximo(descontoVendaComLapisECaneta)
            descontoVendaComLapisECaneta!!.setProximo(semDesconto)
        }
    }

}
